## calcD2stat.py
calculates the D2 and D1 statistics from Hibbins and Hahn 2018

## calcNodeHeights.py
summary of node heights. Requires ete3

## calcNodeHeights.py
summary of node heights. Requires ete3

## calcT1T2.allel.py
calculates the 1st and 2nd node times from Fontaine 2015. Requires scikit-allel

## maf2phy.awk
reformats MAF to phylip

## maf2raxml.py
reformats MAF to run in raxml

## makeUltrametric.py
reformats tree to be ultrameteric

## parseDistmat.py
parse a distmatrix for plotting

## pruneTips.py
prunes tips so only 1 per species. Requires ete3

## traits4network.py
formats a traits file for popart

## trees2reduce.py
a bunch of functions for phylogenetic trees

## tree_summ.py
calculates branch lengths

## BppScrape.py
summarizes MAP from BPP output for plotting in R

## clustPhy.py
clusters phylip formatted files where each file contains 1 locus.

## makeBPPFile.py
creates input files for use with BPP

## makeCTL4BPP.py
make all control file for parallel run of BPP

## makeNegBed.py
creates an inverse bedfile. Used to select intergenic and intron using a gff
